### DISCUSSION TITLE
Refactoring: move module member to different module

### DISCUSSION THREAD

**Discussion by ynouri** [microsoft/pylance-release#2620] [Enhancement]
Related existing issues: microsoft/vscode-python#4628

Same feature as described on this PyCharm documentation page:
https://www.jetbrains.com/help/pycharm/move-refactorings.html#move_module_members

Would be used as such:
* Select variable or function `foo`
* Press shortcut or select "Move" action
* Select an existing destination module `bar.py`
* The imports of `foo` across the whole project are updated to reflect the change

I'm creating a new feature request since microsoft/vscode-python#4628 seems to focus on the file renaming refactoring feature.

**Comment by karthiknadig** [comment]
Thank you for the suggestion! We have marked this issue as "needs decision" to make sure we have a conversation about your idea. We plan to leave this feature request open for at least a month to see how many 👍 votes the opening comment gets to help us make our decision.

**Comment by hmoravec** [comment]
I consider [Move Refactorings](https://www.jetbrains.com/help/pycharm/move-refactorings.html) (and [Changing Signature](https://www.jetbrains.com/help/pycharm/change-signature.html)) a fundamental functionality for every developer working on any non-trivial codebase. I've been a long term PyCharm user but I always give a try to VS Code once in a while to see if it is ready to be my primary IDE. Unfortunately lack of this sends me again back to PyCharm.

I'm also not sure if votes on this issue are the right measure of its importance. I think many developers who give VS Code a try but then are not able to do basic stuff which they are used to do easily in other IDEs like PyCharm just give up and go back to their IDE without digging deep into issues on GitHub in multiple different repos. It took me quite a while to land on this page.

**Comment by douglasrizzo** [comment]
I also consider this a very useful feature. I can't really list the use cases of this except for when I realize that a method/function that is already working properly would be better suited residing in another class/module, but changing all of its usages in the code would be really tedious and error-prone.

Right now I do it by using the `Ctrl + Shift + H` functionality (Replace in Project), but having a foolproof, automated way would be awesome.

**Comment by cperezabo** [comment]
I believe people at Microsoft don't use VSCode at all... It is just impossible to be productive without refactorings, good auto complete, real auto import, and many other things. If you work without that stuff, dude.... download some pro tools and try them off to get inspired!

**Comment by disrupted** [comment]
> I believe people at Microsoft don't use VSCode at all... It is just impossible to be productive without refactorings, good auto complete, real auto import, and many other things. If you work without that stuff, dude.... download some pro tools and try them off to get inspired!

not sure what you're on about, but Pylance has most of the stuff you mentioned

**Comment by cperezabo** [comment]
Hi @disrupted, do you call this real auto import?

https://user-images.githubusercontent.com/3604722/137166146-e7ed9d5c-40fa-4877-b923-94688fcce43a.mp4

It works in some cases where related stuff gets indexed, other way it won't work.

![CleanShot 2021-10-13 at 12 40 30@2x](https://user-images.githubusercontent.com/3604722/137167788-6294ecfc-d007-4d20-a726-04b14f918939.png)

**Comment by jakebailey** [comment]
Can you please try enabling indexing with this?

```
"python.analysis.indexing": true
```

For user code, see also #1055.

**Comment by hmoravec** [comment]
@jakebailey I'm experiencing the same behavior as @cperezabo and I have enabled the indexing option long time ago (although when I checked it now, it says `Unknow Configuration Setting`).

When I use some object from my workspace, at first I have to use `Go to symbol in workspace` for this new symbol and wait until it shows the symbol in dropdown menu, which probably updates the index, then wait a few seconds and then it usually works. But in total this procedure takes longer than if I just wrote the import statement myself :(.

**Comment by jakebailey** [comment]
For auto-import completions from symbols in your workspace, that is explicitly #1055; I would comment there. It's expected at the moment for auto-imports from the workspace to work only on quick fixes, though we'd like to change that, which is what #1055 is all about.

This thread is about refactorings; I'd rather we continue conversations about auto imports in the right issue.

**Comment by cperezabo** [comment]
> Can you please try enabling indexing with this?
> 
> ```
> "python.analysis.indexing": true
> ```
> 
> For user code, see also #1055.

@jakebailey, that doesn't work, I already tried it some time ago, at least not in the expected way.

**Comment by bjarkemoensted** [comment]
Any updates on this? Not being able to refactor is a huge obstacle.

**Reply by ivanlen** [reply]
Yeah, this is a basic refactor functionality. Especially when you are used to pycharm refactor tools. It's kind of downgrading lots of levels :S .

**Comment by heejaechang** [comment]
@bjarkemoensted we started doing incremental rollout for module rename refactoring. member move is being worked on. still require more work.

**Reply by ivanlen** [reply]
any update? thanks :)

**Comment by heejaechang** [comment]
@cperezabo @hmoravec if you can open your own issue with issues you are having, it would be easier for us to track and address the issues. and leave this issue for the refactoring.

**Comment by amir-abdi** [comment]
Any updates?
Is there a release (or unstable/nightly-build release?) scheduled with this feature?

**Comment by emirkmo** [comment]
Seems like this is already available in the JS and TypeScript extensions of VSCode, maybe that could be looked to for help?

This is one of the biggest current downsides to switching from `pycharm` for me. Makes code refactoring a pain.

**Comment by alexmolas** [comment]
Is there any update on this topic? I don't understand why this is not yet implemented. It's evident that the need of this feature is critical for all the developers, but it seems that there's no work towards solving that 😢

**Comment by heejaechang** [comment]
it was blocked by some infrastructure work. about 70% done. work will be resumed soon.

**Comment by cmobarry** [comment]
Recent news on Python move (between modules) refactoring?

**Reply by heejaechang** [reply]
working on it this sprint.

**Reply by DanielMS93** [reply]
will you let us know on this thread when it is complete?

**Reply by heejaechang** [reply]
will do. almost done. but we probably will start with an option that is off by default first.

**Reply by emirkmo** [reply]
@heejaechang, please let us know if you want tests or feedback. I have oodles of long python files to explode out.

**Comment by luabud** [comment]
Hey folks, this is now available in the latest [pre-release version ](https://devblogs.microsoft.com/python/python-in-visual-studio-code-june-2022-release/#pre-release-pylance-extension)of the Pylance extension. You can set `"python.analysis.refactoring.allowMovingSymbols": true` in your settings.json to try it out. This setting is on by default for those using the pre-release version of Pylance in [VS Code Insiders.](https://code.visualstudio.com/insiders/)  Feedback is appreciated 😊

**Reply by MartinBernstorff** [reply]
I cannot express how excited I am about this! I've had `pycharm` installed only so that I could refactor at the module level. Not having to switch to pycharm is a _huge_ quality of life improvement for me, thank you so much! Please pass it along to whomever worked on this! ❤️

**Reply by vbrozik** [reply]
> Thanks, but how should this work?
> 
> I upgraded Pylance to the current pre-release: 2023.7.42...

**Edit:** OK, I have found that the feature was already released in April in the production version but I did not know how to use it. Before starting the **Refactor** action you have to select just the symbol identifier (not the whole class or function) or just move the cursor to the identifier.

More information:
https://devblogs.microsoft.com/python/python-in-visual-studio-code-april-2023-release/#move-symbol-refactoring

---

I also noticed that the feature cannot move symbols created by assignment. Is it planned to add this missing functionality? Example:

``` python
RuleAddressSpec: TypeAlias = IPv4Network | ASAGroupNetwork | IPv4Set

ResourceT = TypeVar('ResourceT', RuleAddressSpec, ASAGroupService)
```

Also moving multiple symbols inside a selection could be useful too.

**Reply by debonte** [reply]
@vbrozik, please file issues if you'd like to see new/different behavior with the move refactoring.

**Reply by noklam** [reply]
Is this feature released already? I have try different ways, i.e. Quick Fix, code action but I cannot trigger the "move symbol" to action.

I see this available in the shortkey settings, but again I fail to trigger it.
![image](https://github.com/microsoft/pylance-release/assets/18221871/efb150aa-3d1d-401b-a5e1-9aa0c05c07bc)

(I am on the latest Pylance & Python extension)

**Reply by vbrozik** [reply]
Yes it was released in April 2023 already. Most often I invoke the function using the "Code Actions" Ctrl+.
![image](https://github.com/microsoft/pylance-release/assets/5073939/5f0b0da3-f102-43cb-be7b-26d1d5d5c79e)

Just move your cursor to be inside the symbol name in the definition and press Ctrl+.

**Reply by noklam** [reply]
When I invoke the command I always see the rewrite section but not Move To.

**Reply by heejaechang** [reply]
You probably didn't see the code action item because what you are selecting is not safely movable, or at least, not with our current implementation. It would be nice if you could open an issue with reproduction steps so we can understand why this is happening and how we can indicate the reason we can't move the symbol in the code action UI somehow.

**Reply by noklam** [reply]
Hey @heejaechang, thanks for jumping in. It would be nice if there is a reason to see why "move symbol" is not available, it would be better to be able to click on it and show an error rather than just making that option disappear. IMO this experience is much smoother in PyCharm, thus I am very excited when I find out VSCode actually support this.

For the reproducer, there is nothing special. I am working on this project, what I have is an editable install of the kedro project. Things like go to definition and find reference works, so I guess the index is working. 

```bash
git clone https://github.com/kedro-org/kedro.git
pip install -e .
```

**Reply by noklam** [reply]
<img width="633" alt="image" src="https://github.com/microsoft/pylance-release/assets/18221871/e517ed32-3e04-4ca8-9794-b8a2c70b6a06">

I just found out that I have an outdate setting for Pylance, before this the repository was in a "Github" folder, after updating the path I can see the "move symbol to" option! I still think this is not ideal because when I do `Ctrl + click`, it goes to the correct file, only this particular feature is failing. It would give me more hint if there is an explicit error.

**Reply by heejaechang** [reply]
move symbol only works between user files, if you install a repo as `editable install` and reference it from `project` (the one you installed repro as editable install), symbols in the repro won't be movable. we will treat those repro as third party package and consider it as `readonly`.

if it is not the case, It would be better to open an issue and provide some context.

**Reply by ShravanSunder** [reply]
![2024-05-15 0548 Brave Browser pylance-releaseCHANGELOG md at main · microsoftpylance-release](https://github.com/microsoft/pylance-release/assets/5294949/d32486d3-7cb6-4a98-9ea0-73becf0e4053)


it doesn't seem like we can move classes to a new file.  I would think that would be an appropriate usecase?

**Reply by heejaechang** [reply]
put cursor on the name of the class rather than select the whole class.

**Comment by pradyunsg** [comment]
I've done a few manual attempts of the basic cases and the pre-release seems to work pretty well on moving a single thing! The import rewriting is appreciated as well. It doesn't clean up the old imports every time, but it's something that pylance does flag and is easy-enough to remove with auto-fixes from import cleaning up logic (ALA ruff).

- One issue that I've hit is when moving files and getting the OS file selection box, it only lets me pick folders rather than files, when I choose to "Move to..." -> manually enter/select thingie.
- One improvement/feature/follow-up I'd appreciate with this: being able to moving a function and it's entire "cone" defined within a file into another file.
  I hit this while trying to split a single file that contains CLI logic + datetime math + API calls to GitHub + API calls to Discourse as functions, and moving them to separate files). It isn't clear to me whether that's even possible at the moment! :)

**Reply by debonte** [reply]
> It doesn't clean up the old imports every time...

@pradyunsg, this is coming soon

Could you please open issues on your two bullet points above?

**Reply by heejaechang** [reply]
> One issue that I've hit is when moving files and getting the OS file selection box, it only lets me pick folders rather than files, when I choose to "Move to..." -> manually enter/select thingie.

I can't repro this on windows. if you can provide me repro steps, I will take a look again.

**Reply by heejaechang** [reply]
> One improvement/feature/follow-up I'd appreciate with this: being able to moving a function and it's entire "cone" defined within a file into another file.

I am not quite sure what you meant by this. can you open an issue with repro steps as Erik suggested?

**Reply by pradyunsg** [reply]
(FWIW, filing this issue is on my TODO list -- I'll do so soon)

**Comment by declanoller** [comment]
I can't seem to get this to work. When I right click on the class or function name, click refactor, and then "move symbol to new file", it gives me the error `cannot find a symbol to move` and nothing happens.

Here are my versions:
```
Version: 1.98.1
Commit: 2fc07b811f760549dab9be9d2bedd06c51dfcb9a
Date: 2025-03-10T15:38:08.854Z
Electron: 34.2.0
ElectronBuildId: 11160463
Chromium: 132.0.6834.196
Node.js: 20.18.2
V8: 13.2.152.36-electron.0
OS: Linux x64 6.11.0-17-generic snap
```
python extension:
```
Identifier
ms-python.python
Version
2025.2.0
Last Updated
2025-03-13, 01:32:26
```
pylance extension:
```
Identifier
ms-python.vscode-pylance
Version
2025.3.1
Last Updated
2025-03-07, 17:01:43
```

Am I using this incorrectly somehow? I've tried the same thing in different modules, etc, same result.

**Reply by rchiodo** [reply]
Do you have some example code? It works for me with this example:

```python
def test_function():
   print('test')
```

You click Refactor on the right click menu and then this should pop up:

![image](https://github.com/user-attachments/assets/2c0d9b46-fe87-4e8d-91de-c09cff5b3bd9)

Well maybe minus the copilot stuff if you don't have copilot installed.

**Reply by declanoller** [reply]
Hi, thanks for the reply. Maybe I'm just misunderstanding how the feature is meant to be used? My expectation was that it's like Pycharm's refactor-->move, where I can move a class or function from one file to another, and it'll update all references to it, and imports for it. Is that right?

I'll use your example. I've made two files, `refactor_move_source.py` and `refactor_move_target.py`, in the same dir (`random_scripts/`). `refactor_move_source.py` has:

```python
def test_function():
    print("test")
```

and `refactor_move_target.py` is empty.

Then, I right click on the function name `test_function`, go down to refactor:

![image](https://github.com/user-attachments/assets/196f1c3b-00af-4c36-86b8-9176aa3e7b74)

click:
![image](https://github.com/user-attachments/assets/10e801e4-d10e-47e0-afd5-807e0da77017)

it opens the search bar:

![image](https://github.com/user-attachments/assets/cfeac5e8-28df-40da-b596-f6e505aa8f91)

I start typing `refactor_move_target.py`, the autocomplete shows it:
![image](https://github.com/user-attachments/assets/3f111234-b92d-4b09-8733-40758a7448ea)

I click on it, and... oh good lord it works now :facepalm: :
![image](https://github.com/user-attachments/assets/c3667107-4b06-413a-9c08-dc20c1c89990)

Okay, but when I go back to my original file I wanted to use this for, it's **still not working!**

It looks like it's about the directory I'm doing it in... I repeated the same process above with the simple `test_function()`, with two files in my original project's directory. So it's not involving my code or their files, it's simply in the same directory as my code. So I have:
![image](https://github.com/user-attachments/assets/37f822d1-be79-425f-975c-71aaa5ebff2e)

and when I repeat the exact process above, after selecting `test_target.py`, I get this error at the bottom right:
![image](https://github.com/user-attachments/assets/61520466-861f-4110-b673-84223c4ddc61)

Do you have any ideas? it's very puzzling... I don't think there's anything special about this directory.

**Reply by declanoller** [reply]
Wait, I think I might've found it!

I made a whole new dir, `random_stuff/`, with the two files above:
![image](https://github.com/user-attachments/assets/2b417292-49c1-4243-8a86-8a2f738f10cd)

If I do the refactor/move process, it works.

Now, if I reset it (so `test_source.py` has `test_function` and `test_target.py` is empty again), but rename the dir to `random-stuff/`, and try again...
![image](https://github.com/user-attachments/assets/a5ad3204-4be5-46b3-98da-2633e36f586c)
![image](https://github.com/user-attachments/assets/c5d9abbe-928e-47e8-b8fa-5e0457b37e04)

womp womp. So it looks like something is tripping it up with the hyphen in the dir name??

Note that my project dir above had a hyphen (`probabalistic-modeling`), while the working example in the other dir didn't (only underscores).

**Reply by rchiodo** [reply]
Let me try with a hyphen

**Reply by rchiodo** [reply]
Yeah I can reproduce the error. I'll enter an issue.

**Reply by rchiodo** [reply]
Thanks for the help finding the bug. I entered it here:
https://github.com/microsoft/pylance-release/issues/7042

**Reply by declanoller** [reply]
I'm surprised no one hit this before, because it's standard for git repos to have hyphens rather than spaces :thinking: 

of course, and thank you for the speedy responses and maintaining it!

**Reply by vbrozik** [reply]
> it's standard for git repos to have hyphens

Yes, by convention repositories have hyphens as separators in their names but Python identifiers (like module names) cannot normally use hyphens, because this character is the minus operator. Identifiers in Python and many other programming languages use underscores as separators.

The most common directory structure of a repository with a Python package looks like this:

``` text
repository-name/
	src/
		main_python_module_name/
			__init__.py
			submodule1.py
			submodule2.py
			submodule3/
				__init__.py
				submodule31.py
				submodule32.py
	pyproject.toml
```

The Python package name is defined inside `pyproject.toml`. In simple cases it would be the same as the `main_python_module_name` and the repository name could also be the same with underscores replaced by hyphens.

**Reply by declanoller** [reply]
@vbrozik that's true, but does that solve the problem here? 

For example, I cloned the [sphinx-pyproject](https://github.com/sphinx-toolbox/sphinx-pyproject) repo (note the hyphen). In its `pyproject.toml` it says `package = "sphinx_pyproject"` and `package_root = "sphinx_pyproject"`. So I put those two `test_source.py` and `test_target.py` files in the dir `sphinx-pyproject/sphinx_pyproject/`, and tried to do the refactor move. I still get the error as above. Maybe I'd have to actually `pip install` it to have that register and get moved correctly?
